cloud/wfm-rest-api/src/impl/MongoDbRepository.ts

```diff
@@ -1,76 +1,74 @@
-import * as Promise from 'bluebird';
-import { Cursor, Db } from 'mongodb';
+import * as Bluebird from 'bluebird';
+import { Collection, Cursor, CursorCommentOptions, Db } from 'mongodb';
 import { ApiError } from '../data-api/ApiError';
 import { defaultPaginationEngine } from '../data-api/MongoPaginationEngine';
 import { DIRECTION, SortedPageRequest } from '../data-api/PageRequest';
 import { PageResponse } from '../data-api/PageResponse';
 import { PagingDataRepository } from '../data-api/PagingDataRepository';
 import * as errorCodes from './ErrorCodes';
 
-const dbError: ApiError = { code: errorCodes.DB_ERROR, message: 'MongoDbRepository database not intialized' };
+const dbError: ApiError = new ApiError(errorCodes.DB_ERROR, 'MongoDbRepository database not intialized');
 
 /**
  * Service for performing data operations on mongodb database
  */
-export class MongoDbRepository implements PagingDataRepository {
+export class MongoDbRepository<T extends { id: string }> implements PagingDataRepository<T> {
 
-  public db: any;
+  public db: Db;
+  protected collection: Collection<T>;
 
   /**
    * @param collectionName - name of the collection stored in mongodb
    */
   constructor(readonly collectionName: string) {
-    const self = this;
   }
 
-  public list(filter: any, request: SortedPageRequest): Promise<PageResponse> {
+  public list(filter: object | CursorCommentOptions, request: SortedPageRequest): Bluebird<PageResponse<T>> {
     if (!this.db) {
-      return Promise.reject(dbError);
+      return Bluebird.reject(dbError);
     }
-    const cursor: Cursor = this.db.collection(this.collectionName).find(filter);
-    return new Promise((resolve, reject) => {
-      cursor.count(filter, function(err, totalNumber) {
-        if (err) {
-          return reject(err);
-        }
-        return resolve(defaultPaginationEngine.buildPageResponse(request, cursor, totalNumber));
-      });
-    });
+    const cursor = this.collection.find(filter);
+    return Bluebird.resolve(cursor.count(true, filter))
+      .then(count => defaultPaginationEngine.buildPageResponse(request, cursor, count));
   }
 
-  public get(id: string) {
+  public get(id: string): Bluebird<T> {
     if (!this.db) {
-      return Promise.reject(dbError);
+      return Bluebird.reject(dbError);
     }
-    return this.db.collection(this.collectionName).findOne({ id });
+    return Bluebird.resolve(this.collection.findOne({ id }));
```

Comment:
[Question] Why we wrapping this call. I should return promise anyway? Good pattern?